### PR TITLE
Fits nucal models to data

### PR DESCRIPTION
**Pull request by tyler-a-cox**
This PR adds new functions for performing linear fits of nucal modes to data. Implements methods which fit u-dependent models and full u-nu dependent models to the data to find a decent starting point for gradient descent. Full gradient descent fitting will be implemented in the next `nucal` PR

### CODE REVIEW

File: hera_cal/nucal.py [hunk 1]
```diff
@@ -507,21 +578,13 @@ def compute_spatial_filters(radial_reds, freqs, spatial_filter_half_width=1, eig
...
-
     # Loop through each baseline in each radial group
-    for gi, group in enumerate(radial_reds):
-        umin, umax = u_bounds[gi]
-        for bl in group:
-            umodes = radial_reds.baseline_lengths[bl] / SPEED_OF_LIGHT * freqs
-            pswf, _ = dspec.pswf_operator(
-                umodes, filter_centers=[0], filter_half_widths=[spatial_filter_half_width], eigenval_cutoff=[eigenval_cutoff], 
-                xmin=umin, xmax=umax, cache=cache
+    for group in radial_reds:
```
Comment (jsdillon) [inline hera_cal/nucal.py:582]: Is this inefficient? Is there a circumstance where you'd be computing the same filters for different orientations? Or do we not really care?

File: hera_cal/nucal.py [hunk 2]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
         data_is_redsol=data_is_redsol, gain_flags=gain_flags, tol=tol, antpos=antpos
     )
 
-    return wgts
\ No newline at end of file
+    return wgts
+
+def _linear_fit(XTX, Xy, solver='lu_solve', tol=1e-15):
```
Comment (jsdillon) [inline hera_cal/nucal.py:689]: This feels like it probably belongs in hera_filters, rather than here, but I'm fine including it here for now if you want to make an issue to move it.

File: hera_cal/nucal.py [hunk 3]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+    return model
+
+
+def fit_nucal_foreground_model(data, data_wgts, radial_reds, spatial_filters, spectral_filters=None, tol=1e-15,
+                               share_fg_model=False, return_model_comps=False, solver="lu_solve"):
+    """
+    Compute a foreground model for a set of radially redundant baselines. The model is computed by performing a linear
+    least-squares fit using a set of DPSS filters to model visibilities within a radially redundant group. If only spatial
+    filters are provided, the model is restricted to the spatial axis. If spectral filters are also provided, the model is fit
+    allowing spectral variation at a fixed spatial scale. The model components are return if return_model_comps is True.
```
Comment (jsdillon) [inline hera_cal/nucal.py:819]: are return: typo

File: hera_cal/nucal.py [hunk 4]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+    least-squares fit using a set of DPSS filters to model visibilities within a radially redundant group. If only spatial
+    filters are provided, the model is restricted to the spatial axis. If spectral filters are also provided, the model is fit
+    allowing spectral variation at a fixed spatial scale. The model components are return if return_model_comps is True. 
+    Otherwise, the model is evaluated using the model components and the spatial filters (and spectral filters if provided),
+    and the model visibilities are returned.
+    
+    Parameters:
+    -----------
+    data : dictionary, DataContainer
+        Data to be fit by u-model. This function assumes that visibilities redundantly-averaged
```
Comment (jsdillon) [inline hera_cal/nucal.py:826]: visibilities redundantly-averaged: typo

File: hera_cal/nucal.py [hunk 5]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+        and will only use data from the first baseline in a spatially-redundant group.
+    data_wgts : dictionary, DataContainer
+        Weights associated with data
+    radial_reds : RadialRedundancy, list of lists of tuples
+        List of lists of redundant baseline tuples. Each list of redundant tuples represents a radial group of
```
Comment (jsdillon) [inline hera_cal/nucal.py:831]: I'd clarify where you mean baseline redundancy and where you mean frequency redundancy

File: hera_cal/nucal.py [hunk 6]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+    # Loop over radial groups
+    for group in radial_reds:
+        # Get the data and design matrix for this group
+        design_matrix = np.array([spatial_filters[bl] for bl in group])
+        wgts_here = np.array([data_wgts[bl] for bl in group])
+        data_here = np.array([data[bl] for bl in group])
+                
+        # Get number of fit parameters
+        if spectral_filters is not None:
+            ndim = spectral_filters.shape[1] * design_matrix.shape[-1]
```
Comment (jsdillon) [inline hera_cal/nucal.py:874]: I'd add a comment explaining which dimensions these indices are

File: hera_cal/nucal.py [hunk 7]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+
+        # Compute the XTX
+        if share_fg_model:
+            # Compute the XTX and XTWy
+            if spectral_filters is None:
+                XTX = jnp.einsum("afm,atf,afn->mn", design_matrix, wgts_here, design_matrix)
```
Comment (jsdillon) [inline hera_cal/nucal.py:880]: These einsums can be a bit hard to fllow. I'd suggest commenting which dimension corresponds to which index letter

File: hera_cal/nucal.py [hunk 8]
```diff
@@ -621,4 +684,281 @@ def build_nucal_wgts(data_flags, data_nsamples, autocorrs, auto_flags, radial_re
...
+    spectral_filters : np.ndarray, optional, Default is None.
+        Array of DPSS filters with shape (Nfreqs, Nspec) where Nspec is the number of spectral DPSS filters. If None,
+        the model is assumed to be restricted to the spatial axis with no spectral variation.
+    solver : str, optional, Default is 'lu_solve'
+        Solver to use for linear least-squares fit. Options are 'lu_solve', 'solve', 'pinv', and 'lstsq'.
+    tol : float, optional, Default is 1e-15.
+        Regularization tolerance for linear least-squares fit. 
+    share_fg_model : bool, optional, Default is False.
+        If True, the foreground model for each radially-redundant group is shared across the time axis. 
+        Otherwise, a nucal foreground will be independently computed for each frequency.
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/nucal.py:845]: Also, if I'm understanding this correctly, this means that all times are considered independent samples that you're fitting a single model to? Perhaps you might want to clarify that in the docstring?

[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:845]: Oops, yeah that’s suppose to be time not frequency. And yes, by default it fits a model to each time independently, but you can fit a model jointly across time, assuming the sky doesn’t change much across a couple time samples or the time axis of your data array are different days at a particular LST. I’ll make that more clear in the docstring

File: hera_cal/nucal.py [hunk 9]
```diff
@@ -477,9 +481,76 @@ def sort(self, key=None, reverse=True):
...
+        Fourier half width of the spectral filter in units of seconds
+    eigenval_cutoff : float, optional, default=1e-12
+        Cutoff for the eigenvalues of the DPSS filter
+
+    Returns:
+    -------
+    spectral_filters : np.ndarray
+        Array of spectral filters with shape (Nfreqs, Nfilters)
+    """
+    return dspec.dpss_operator(freqs, [0], [spectral_filter_half_width], eigenval_cutoff=[eigenval_cutoff])[0].real
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:503]: You can phase these DPSS to particular delays so by default they’re complex with zeros in the imaginary component. It’s not totally necessary to do though

File: hera_cal/nucal.py (hunk 1 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:582]: It may be a little, but the filters are dependent on baseline length and the difference between the minimum and maximum u of a particular group, so I was just assuming there would probably only be a few cases where you’d be recomputing the same filters. I have a speed-up that I have in mind for hera_filters that should speed this up by 2-3x anyway, so I don’t think this will be a huge bottleneck.

File: hera_cal/nucal.py [hunk 10]
```diff
@@ -477,9 +481,76 @@ def sort(self, key=None, reverse=True):
...
+    Compute prolate spheroidal wave function (PSWF) filters for a single radially redundant group.
+
+    Parameters:
+    ----------
+    group : list
+        List of baseline tuples that are radially redundant
+    freqs : np.ndarray
+        Array of frequencies in Hz
+    bls_lengths : dict
+        Dictionary of baseline lengths in meters
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:517]: Good point I’ll make that change

File: hera_cal/nucal.py (hunk 7 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:880]: Sounds good, I’ll make sure to do that.

File: hera_cal/nucal.py (hunk 3 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:819]: Fixed this

File: hera_cal/nucal.py (hunk 5 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:831]: Fixed this

File: hera_cal/nucal.py (hunk 6 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:874]: Added to the comment above

File: hera_cal/nucal.py (hunk 4 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:826]: Fixed

File: hera_cal/nucal.py (hunk 2 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:689]: Will do. I'll make an issue after this is merged in

**Review by jsdillon — APPROVED** [review]
Looks good! Don't forget to make the issue to move _linear_fit to hera_filters
